File: python/modifying_sequence.py
# Implementations of modifying sequence algorithms
from itertools import islice, ifilter, ifilterfalse, imap
from random import randint

def copy(iter, start, end):
    return islice(iter, start, end)

def copy_if(iter, func):
    return ifilter(func, iter)

def copy_n(iter, count):
    return islice(iter, 0, count)

# require __slice__ is implemented
def copy_backward(iter, start, end):
    return reversed(iter[start:end])

# there are not move semantics in python(unlike C++), so implement it as copy
def move(iter, start, end):
    return islice(iter, start, end)

def move_backward(iter, start, end):
    return reversed(iter[start:end])

def fill(seq, value):
    for i in range(0, len(seq)) :
        seq[i] = value

def fill_n(seq, value, count):
    count = min(count, len(seq))
    for i in range(0, count) :
        seq[i] = value

def transform(iter, func):
    return [func(i) for i in iter]

def generate(seq, func):
    for i in range(0, len(seq)) :
        seq[i] = func()

def generate_n(seq, func, count):
    for i in range(0, count) :
        seq[i] = func()

# the problem is, iterate in dict return (k1, k2, ...) instead of ((k1, v1), ...)
def remove(iter, value):
    return ifilterfalse(lambda x: x == value, iter)

def remove_if(iter, func):
    return ifilterfalse(func, iter)

def remove_copy(iter, value):
    return [i for i in remove(iter, value)]

def remove_copy_if(iter, func):
    return [i for i in remove_if(iter, func)]

def replace(iter, before, after):
    def _replace_func(x):
        if x == before :
            return after
        else :
            return x
    return imap(_replace_func, iter)

def replace_if(iter, func, after):
    def _replace_if(x):
        if func(x) :
            return after
        else :
            return x
    return imap(_replace_if, iter)

def replace_copy(iter, before, after):
    return [i for i in replace(iter, before, after)]

def replace_copy_if(iter, func, after):
    return [i for i in replace_if(iter, func, after)]

def swap(a, b):
    return b, a

def swap_range(a, b, start, end):
    for i in range(start, end) :
        a[i], b[i] = b[i], a[i]

def iter_swap(a, b):
    return b, a

# require __setitem__ is implemented
def reverse(seq):
    end = len(seq) - 1
    for i in range(0, end / 2) :
        seq[i], seq[end - i] = seq[end - i], seq[i]

def reverse_copy(iter):
    return [i for i in reversed(iter)]

# require extra space to store tmp.
# An algorithm without extra space is possible, but not elegant.
def rotate(seq, pivot):
    tmp = seq[:pivot]
    tail = len(seq) - pivot
    for i in range(0, len(seq)) :
        if i < tail : seq[i] = seq[pivot + i]
        else : seq[i] = tmp[i - tail]

def rotate_copy(seq, pivot):
    return seq[pivot:] + seq[:pivot]

def shuffle(seq):
    length = len(seq)
    for i in range(0, length) :
        random = randint(0, i)
        seq[i], seq[random] = seq[random], seq[i]

def unique(seq):
    current = None
    for i in range(0, len(seq)) :
        if current != seq[i] :
            current = seq[i]
            yield seq[i]

def unique_copy(seq):
    return [i for i in unique(seq)]

File: php/ModifyingSequence.php
<?php
namespace Algorithm;

function copy(array $ary, $start, $end) {
    return array_slice($ary, $start, $end - $start);
}

function copyIf(array $ary, $func) {
    $res = [];
    foreach ($ary as $e)
        if ($func($e))
            array_push($res, $e);
    return $res;
}

function copyN(array $ary, $n, $value) {
    return generateN($ary, $n, function() use ($value) {
        return $value;
    });
}

function copyBackward(array &$dest, array $src) {
    $bound = max(0, \count($dest)-\count($src));
    for ($i = \count($dest)-1, $j = \count($src)-1; $i >= $bound; $i--, $j--)
        $dest[$i] = $src[$j];
}

function fill(array $ary, $value) {
    return array_pad([], \count($ary), $value);
}

function fillN(array $ary, $n, $value) {
    if ($n < \count($ary))
        return copyN($ary, $n, $value);
    else
        return array_pad([], $n, $value);
}

function transform(array $ary, $func) {
    $res = $ary;
    foreach ($res as $k => $v)
        $res[$k] = $func($v);
    return $res;
}

function generate(array $ary, $func) {
    $res = $ary;
    foreach ($res as $k => $v)
        $res[$k] = $func();
    return $res;
}

function generateN(array $ary, $n, $func) {
    $res = $ary;
    $bound = min(\count($ary), $n);
    for ($i = 0; $i < $bound; $i++)
            $res[$i] = $func();
    return $res;
}

function remove(array &$ary, $value) {
    removeIf($ary, function($x) use ($value) {
        return $x === $value;
    });
}

function removeIf(array &$ary, $func) {
    foreach ($ary as $k => $v)
        if ($func($v))
            unset($ary[$k]);
    $ary = array_values($ary);
}

function removeCopy(array &$ary, $value) {
    return removeCopyIf($ary, function($x) use ($value) {
        return $x === $value;
    });
}

function removeCopyIf(array &$ary, $func) {
    $res = [];
    foreach ($ary as $k => $v)
        if ($func($v)) {
            array_push($res, $v);
            unset($ary[$k]);
        }
    $ary = array_values($ary);
    return $res;
}

function replace(array &$ary, $before, $after) {
    replaceIf($ary, function($x) use ($before) {
        return $x === $before;
    }, $after);
}

function replaceIf(array &$ary, $func, $after) {
    foreach ($ary as $k => $v)
        if ($func($v))
            $ary[$k] = $after;
}

function replaceCopy(array $ary, $before, $after) {
    return replaceCopyIf($ary, function($x) use ($before) {
        return $x === $before;
    }, $after);
}

function replaceCopyIf(array $ary, $func, $after) {
    $res = [];
    foreach ($ary as $k => $v)
        array_push($res, $func($v) ? $after : $v);
    return $res;
}

function swap(&$a, &$b) {
    $tmp = $a;
    $a = $b;
    $b = $tmp;
}

function swapRange(array &$a, array &$b, $start, $end) {
    $bound = min(\count($a), \count($b), $end);
    if ($bound <= $start) return;
    for ($i = $start; $i < $bound; $i++)
        swap($a[$i], $b[$i]);
}

function _reverse(array &$ary, $start, $end) {
    $end--;
    for ($i = $start; $i < ($end+$start) / 2; $i++)
        swap($ary[$i], $ary[$end+$start-$i]);
}

function reverse(array &$ary) {
    $length = \count($ary);
    _reverse($ary, 0, $length);
}

function reverseCopy(array $ary) {
    $res = [];
    for ($i = \count($ary) - 1; $i >= 0; $i--)
        array_push($res, $ary[$i]);
    return $res;
}

function shuffle(array &$ary) {
    for ($i = 0; $i < \count($ary); $i++) {
        $rd = rand(0, $i);
        swap($ary[$i], $ary[$rd]);
    }
}

function correctPivot($pivot, $length) {
    $pivot = $pivot % $length;
    if ($pivot < 0) $pivot += $length;
    return $pivot;
}

function rotate(array &$ary, $pivot) {
    $length = \count($ary);
    if ($length === 0) return $ary;
    $pivot = correctPivot($pivot, $length);
    _reverse($ary, 0, $pivot);
    _reverse($ary, $pivot, $length);
    _reverse($ary, 0, $length);
}

function rotateCopy(array $ary, $pivot) {
    $length = \count($ary);
    if ($length === 0) return $ary;
    $pivot = correctPivot($pivot, $length);
    return array_merge(array_slice($ary, $pivot),
                       array_slice($ary, 0, $pivot));
}

function unique(array &$ary) {
    $set = [];
    foreach ($ary as $k => $v)
        if (in_array($v, $set))
            unset($ary[$k]);
        else
            array_push($set, $v);
    $ary = $set;
}

function uniqueCopy(array $ary) {
    $res = [];
    foreach ($ary as $e)
        if (!in_array($e, $res))
            array_push($res, $e);
    return $res;
}
?>
